### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by asish918**
# Homework 4 Asish

### Topics Covered:
- Exceptions/Chaining
- Comparator/Iterator
- Collections

### Drift Video Explanation
[Drift](https://video.drift.com/v/ab0khEh4n89lb1RIJQgi5GxIiwtVM6zymYgzy0YvLk)

### Output Screenshots - 
Screenshot 2024-01-12 at 01 36 37

Screenshot 2024-01-12 at 01 36 23

Screenshot 2024-01-12 at 01 36 10

Screenshot 2024-01-12 at 01 35 52

### CODE REVIEW

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 0 |
| code_smells | 0 |
| reliability_rating | A (0 Bugs)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/homework/2/src/main/java/asish/kdu/q1/StudentUtil.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    public static int[] getStudentsByGPA(double lower, double higher, int[]
+            studentIdList, char[][] studentsGrades) {
+        if(lower < 0 || higher < 0 || lower > 4 || higher > 4 || higher < lower)
+            return new int[0];
+
+        double[] gpaResult = calculateGPA(studentIdList, studentsGrades);
+
+        int resultSize = 0;
+
+        for (double v : gpaResult) {
```
Comment (aakash232) [inline backend/homework/2/src/main/java/asish/kdu/q1/StudentUtil.java:45]: v can be replaced with a more meaningful variable name

File: backend/homework/2/src/main/java/asish/kdu/logging/CustomLogger.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CustomLogger {
+    private CustomLogger() {
+
+    }
+    private static final Logger logger
+            = LoggerFactory.getLogger(CustomLogger.class);
+    public static void printLogger(String message) {
```
Comment (aakash232) [inline backend/homework/2/src/main/java/asish/kdu/logging/CustomLogger.java:12]: More logger levels could be exposed as methods and used throughout

File: backend/homework/2/src/main/java/asish/kdu/q1/Q1Utils.java
Comment (aakash232) [inline backend/homework/2/src/main/java/asish/kdu/q1/Q1Utils.java:1]: Let class names be meaningful to the logic implemented.
Package names can denote question numbers

File: backend/homework/2/src/main/java/asish/kdu/q2/SentimentAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            if (opinion == 0)
+                opinion = Q2Utils.firstPatternOpinionLookup(sentence, negOpinionWords, feature, Q2Utils.Opinion.NEGATIVE);
+
+            if (opinion != 0) return opinion;
+        }
+
+        return 0;
+    }
+
+    public static void main(String[] args) {
```
Comment (aakash232) [inline backend/homework/2/src/main/java/asish/kdu/q2/SentimentAnalyzer.java:64]: General comment throughout : Add javadocs and documentation

**Pull request closed without merging**